### DISCUSSION TITLE
hydra.searchpath vs --config-dir

### DISCUSSION THREAD

**Discussion by jchia** [facebookresearch/hydra#2270] [Q&A]
I found that on the command-line `--config-dir /some/dir` is not the same as `hydra.searchpath=[file:///some/dir]`.

If there is a top-level config file `/some/dir/hyd.yml` and my app uses `@hydra.main(version_base='1.2', config_name='hyd')`, the `--config-dir` works and the `hydra.searchpath` fails. It fails to find the config file with a `hydra.errors.MissingConfigException`.

A superficial reading of https://hydra.cc/docs/advanced/search_path/ suggests that `hydra.searchpath=[file:///some/dir]` should work. What am I missing?

**Comment by Jasha10** [comment]
I can reproduce the behavior you've described.
I believe this fails because overriding `hydra.searchpath` (either at the command line or within your primary config) will only affect discovery of non-primary configs. Looking at Hydra's implementation, it seems to me that the primary config must be in the directory indicated by `--config-dir` or by the `config_path` argument to `hydra.main`.
```python
@hydra.main(version_base='1.2', config_name='hyd', config_path='.')
```
The searchpath is then computed based on what is found in the primary config. This searchpath may be overridden at the CLI, but the primary config is already loaded (or, in your case, failed to load) by the time that the CLI override is processed.

**Reply by jchia** [reply]
Then, it may be helpful for the documentation to say that searchpath does not apply to the primary config. I'm new to Hydra, but I did not get this from reading the current documentation.

Would a PR for such a documentation change be welcome?

**Reply by Jasha10** [reply]
Yes, that would be welcome!
